### DISCUSSION TITLE
ASP.NET Core 3.1 - WebApplicationFactory - What is the correct usage for WebApplicationFactoryContentRootAttribute?

### DISCUSSION THREAD

**Discussion by dcs3spp** [dotnet/aspnetcore#27643] [Q&A]
I am learning how to use the WebApplicationFactory in ASP.NET Core 3.1. In particular I am trying to figure out the correct usage of [WebApplicationFactoryContentRootAttribute](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactorycontentrootattribute?view=aspnetcore-3.0)

I am trying to use this attribute to set the content root for a Startup bootstrap class at ```../../../Src/WebApp```. 

When I run ```dotnet test``` for the code below it does not appear to be referencing the *WebApplicationFactoryContentRootAttribute* settings.

A ```System.IO.DirectoryNotFoundException``` is thrown : 'my solution base dir'/WebApp.
The path should be 'my solution base dir'/Src/WebApp.

How do I get the derived [WebApplicationFactory](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactory-1?view=aspnetcore-3.0) to recognise the [WebApplicationFactoryContentRootAttribute](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactorycontentrootattribute?view=aspnetcore-3.0)

**Integration Test**
```C#
using System.IO;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;

using Autofac.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.AspNetCore.Mvc.Testing;
using Xunit;
using Xunit.Abstractions;

using WebApp;
using WebApp.S3.Contracts;
using WebApp.Functional.Utilities;


/// Key should match FullName of assembly containing TStartup : WebApplicationFactory<TStartup> 
/// https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactorycontentrootattribute?view=aspnetcore-3.0
[assembly: WebApplicationFactoryContentRoot(
    "WebApp, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null",
    "../../../Src/WebApp",
    "Program.cs",
    "1")]


namespace WebApp.FunctionalTests
{
    /// <summary>
    /// Eventual Goal: Tryng to create WebApplicationFactory with content root at Src/WebApp using
    /// appsettings file derived from ASPNETCORE_ENVIRONMENT variable. If
    /// the variable is unset then defaults to appsettings.Local.json. 
    /// This will also be run in a Docker container
    /// </summary>
    public class AppTestFixture : WebApplicationFactory<Startup>
    {
        //override methods here as needed for Test purpose
    }


    public class ApiIntegrationTest : IClassFixture<AppTestFixture>
    {
        private WebApplicationFactory<Startup> _factory;
        private ITestOutputHelper _output;


        /// <summary>
        /// Initialise derived WebApplicationFactory and output stream
        /// </summary>
        /// <param name="factory"><see cref="AppTestFixture"/>, dervied WebApplicationFactory</param>
        /// <param name="output">xUnit output stream</param>
        public ApiIntegrationTest(AppTestFixture factory, ITestOutputHelper output)
        {
            const string contentRoot = "Src/WebApp";

            _factory = factory;
            _output = output;

            // sanity check for FullName property of assembly containing Startup class
            string fullName = Assembly.GetAssembly(typeof(Startup)).FullName;
            _output.WriteLine($"FullName property of assembly containing 'Startup' class => {fullName}");
        }

        /// <summary>
        /// /Users/me/Development/dotnet/blazormotiondetectionlistener/WebApp/ not found
        ///
        /// Should be path /Users/me/Development/dotnet/blazormotiondetectionlistener/Src/WebApp/
        /// </summary>
        ///
        /// <remarks>
        /// How do I use WebApplicationFactoryContentRoot attribute to correctly resolve this?
        /// </remarks>
        [Fact]
        public async Task WebApp_App_ApiController_Test()
        {
            var client = _factory.CreateClient();

            await Task.CompletedTask;
        }
    }
}
```

**Test project**
```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>

    <IsPackable>false</IsPackable>
  </PropertyGroup>

  <ItemGroup>
    <Content Update="xunit.runner.json">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </Content>
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="Autofac" Version="6.0.0" />
    <PackageReference Include="Microsoft.AspNetCore" Version="2.2.0" />
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.2.0" />
    <PackageReference Include="xunit" Version="2.4.1" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.1" />
    <PackageReference Include="coverlet.collector" Version="1.0.1" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\..\..\Src\WebApp\WebApp.csproj" />
    <ProjectReference Include="..\..\Functional\Utilities\WebApp.Functional.Utilities.csproj" />
    <ProjectReference Include="..\..\..\Src\WebApp.S3.Contracts\WebApp.S3.Contracts.csproj" />
  </ItemGroup>

</Project>
```

Also tried adding the Assembly attribute in the test csproj file by adding this fragment:
```C#
<Target Name="AddGitMetadaAssemblyAttributes"
          BeforeTargets="CoreGenerateAssemblyInfo">
    <ItemGroup>
        <AssemblyAttribute Include="Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactoryContentRoot">
            <_Parameter1>WebApp, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null</_Parameter1>
            <_Parameter2>..\..\..\Src\WebApp</_Parameter2>
            <_Parameter3>Program.cs</_Parameter3>
            <_Parameter4>1</_Parameter4>
        </AssemblyAttribute>
    </ItemGroup>
  </Target>
```

Can see that this adds the attribute to the generated AssemblyInfo.cs in *obj/Debug/netcoreapp3.1*:
```C#
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

using System;
using System.Reflection;

[assembly: System.Reflection.AssemblyCompanyAttribute("WebApp.FunctionalTests")]
[assembly: System.Reflection.AssemblyConfigurationAttribute("Debug")]
[assembly: System.Reflection.AssemblyFileVersionAttribute("1.0.0.0")]
[assembly: System.Reflection.AssemblyInformationalVersionAttribute("1.0.0")]
[assembly: System.Reflection.AssemblyProductAttribute("WebApp.FunctionalTests")]
[assembly: System.Reflection.AssemblyTitleAttribute("WebApp.FunctionalTests")]
[assembly: System.Reflection.AssemblyVersionAttribute("1.0.0.0")]
[assembly: Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactoryContentRoot("WebApp, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null", "../../../Src/WebApp", "Program.cs", "1")]
```

So, how do I get the .NET Core WebApplicationFactory<TStartup> to look at 'solutiondir'/Src/WebApp instead of 'solutiondir'/WebApp'? Am I using the wrong value for the attribute key property?

According to [Microsoft documentation](https://docs.microsoft.com/ru-ru/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactory-1?view=aspnetcore-3.0#constructors)
> This constructor will infer the application content root path by searching for a WebApplicationFactoryContentRootAttribute on the assembly containing the functional tests with a key equal to the TEntryPoint assembly FullName. In case an attribute with the right key can't be found, WebApplicationFactory<TEntryPoint> will fall back to searching for a solution file (*.sln) and then appending TEntryPoint assembly name to the solution directory. The application root directory will be used to discover views and content files.

I am using a key value set to the FullName of the assembly containing my Startup class. Is this correct?

**Comment by crypticviper** [comment]
I stumbled upon this post while looking for answers on how to make integration tests run within a CI task in azure devops. In my case the issue was that I was executing tests with published tests (so no solution file). Ultimately I got it to work with just the csproj changes as below

`  <Target Name="AddIntegrationTestAttributes"
        BeforeTargets="CoreGenerateAssemblyInfo">
    <ItemGroup>
      <AssemblyAttribute Include="Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactoryContentRoot">
        <_Parameter1>MyWebService.AssemblyName, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null</_Parameter1>
        <_Parameter2>../ArtifactDirectoryName/</_Parameter2>
        <_Parameter3>appsettings.json</_Parameter3>
        <_Parameter4>1</_Parameter4>
      </AssemblyAttribute>
    </ItemGroup>
  </Target>`

I used the appSettings.json file as tracker as mentioned in one of the other issues:

**https://github.com/dotnet/aspnetcore/issues/16978**

**Comment by dcs3spp** [comment]
@baligavinod Are you setting the key value for ```WebApplicationFactoryContentRoot``` to be the FullName of the assembly for TStartup, e.g. WebApplicationFactoryContentRoot<TStartup>? In my case my TStartup class is derived from the live Startup class for testing purposes. The live Startup class exists in a different assembly to TStartup. I think I am encountering an issue because I am using an incorrect value for the ```key``` property of the attribute and so the factory resorts to trying to find solutiondir. If I use the live Startup class it all works.

If I use ```UseSolutionRelativeContentRoot``` it works locally. However it fails in Docker container with an ```InvalidOperationException``` informing that cannot locate solution root.

**Comment by crypticviper** [comment]
key value for WebApplicationFactoryContentRoot is just assembly name in my case. Custom web application factory implementation looks like below

`CustomWebApplicationFactory : WebApplicationFactory<Startup>`

Startup is the startup class living in web application project. I don't have inherited startup class in my case.

**Reply by dcs3spp** [reply]
Cheers thanks for responding @baligavinod. I am resorting to trying to create a small repository to debug my code. Thanks for responding, appreciated 👍
